### PR TITLE
refactor(pipeline): query any match labels only once instead of multiple queries

### PR DESCRIPTION
**Pull request by chengjoey**
#### What this PR does / why we need it:
query any match labels only once instead of multiple queries

#### Which issue(s) this PR fixes:

- Fixes #your-issue_number
- [Erda Cloud Issue Link](https://erda.cloud/erda/dop/projects/387/issues/gantt?filter__urlQuery=eyJhc3NpZ25lZSI6WyIxMDAxMjA1Il19&id=311070&iterationID=1224&pId=0&type=TASK)

#### Specified Reviewers:

/assign @sfwn 

#### ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English | query any match labels only once instead of multiple queries             |
| 🇨🇳 中文    |     减少标签的查询次数         |

#### Need cherry-pick to release versions?

Add comment like `/cherry-pick release/1.0` when this PR is merged.

> For details on the cherry pick process, see the [cherry pick requests](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md#how-to-cherry-pick-a-merged-pr) section under [CONTRIBUTING.md](https://github.com/erda-project/erda/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by chengjoey** [comment]
> any unit test?

non...it's difficult to write unit-test

**Comment by chengjoey** [comment]
added unit-test

File: internal/tools/pipeline/dbclient/op_pipeline.go [hunk 1]
```diff
@@ -267,14 +267,17 @@ func (client *Client) PageListPipelines(req apistructs.PipelinePageListRequest,
 	var wg sync.WaitGroup
 	var errs []string
 
+	// 基础信息表获取到的 pipelineIDs
+	baseSQL := session.Table(&spec.PipelineBase{}).Where("").Cols(tableFieldName((&spec.PipelineBase{}).TableName(), "id"))
```
Comment (sfwn) [inline internal/tools/pipeline/dbclient/op_pipeline.go:271]: meaningless movement?

File: internal/tools/pipeline/dbclient/op_pipeline.go [hunk 2]
```diff
@@ -267,14 +267,17 @@ func (client *Client) PageListPipelines(req apistructs.PipelinePageListRequest,
...
+
 	// labels 获取到的 pipelineIDs
 	var labelPipelineIDs []uint64
 	var needFilterByLabel bool
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		// select by labels
-		if len(req.MustMatchLabels) > 0 || len(req.AnyMatchLabels) > 0 {
+	if len(req.MustMatchLabels) > 0 || len(req.AnyMatchLabels) > 0 {
```
Comment (sfwn) [inline internal/tools/pipeline/dbclient/op_pipeline.go:276]: nothing actually changed

**Review by sfwn — CHANGES_REQUESTED** [review]
nothing changed in `internal/tools/pipeline/dbclient/op_pipeline.go`.
I think you can revert all changed in this file.

File: internal/tools/pipeline/dbclient/op_pipeline_labels.go [hunk 3]
```diff
@@ -196,23 +196,29 @@ func (client *Client) SelectTargetIDsByLabels(req apistructs.TargetIDSelectByLab
 
 	if len(req.AnyMatchLabels) > 0 {
 		var allNeedMergeTargetIDs [][]uint64
+		innerSegments := append(sqlSegments, " AND (")
```
Comment (sfwn) [inline internal/tools/pipeline/dbclient/op_pipeline_labels.go:199]: maybe you should add unit tests for how to assemble `innerSegments`

**Review by sfwn — CHANGES_REQUESTED** [review]

**Comment by sfwn** [comment]
add screenshot of realistic sqls before and after this pr.

File: internal/tools/pipeline/dbclient/op_pipeline.go (hunk 2 above)
Comment (chengjoey) [reply internal/tools/pipeline/dbclient/op_pipeline.go]: Return to the original condition

File: internal/tools/pipeline/dbclient/op_pipeline_labels.go (hunk 3 above)
Comment (chengjoey) [reply internal/tools/pipeline/dbclient/op_pipeline_labels.go:199]: Return to the original condition

**Comment by sfwn** [comment]
> before: image image
> 
> after: image
> 
> Combine two queries into one when query anymatchlabels
> 
> got same result after sql combined

Very clear. How about `MustMatchLabels`? Maybe just change `OR` to `AND`.

**Review by sfwn — APPROVED** [review]

**Comment by chengjoey** [comment]
> > before: image image
> > after: image
> > Combine two queries into one when query anymatchlabels
> > got same result after sql combined
> 
> Very clear. How about `MustMatchLabels`? Maybe just change `OR` to `AND`.

image
`MustMatchLabels` use `And` operation will cause empty result

**Pull request merged**